src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -340,6 +340,207 @@ class AutoDiffWhileFwd
+      }
+      auto condIterVar = cond->getArgument(0);
+
+      OpBuilder condBuilder(cond, cond->end());
+      condBuilder.create<ReturnOp>(
+          orig->getLoc(),
+          ValueRange(condBuilder
+                         .create<CompareOp>(orig->getLoc(), condIterVar,
+                                            numIters, ComparisonDirection::LT)
+                         .getResult()));
+    }
+
+    bool valid = true;
+    {
+      for (auto operand : operands) {
+        body->addArgument(operand.getType(), orig->getLoc());
+      }
+      OpBuilder bodyBuilder(body, body->end());
+      auto one = bodyBuilder.create<ConstantOp>(
+          orig->getLoc(), unrankedTensorType,
+          SplatElementsAttr::get(unrankedTensorType,
+                                 ArrayRef<Attribute>(IntegerAttr::get(
+                                     bodyBuilder.getI64Type(), 1))));
+      Value bodyIterVar =
+          bodyBuilder.create<AddOp>(orig->getLoc(), body->getArgument(0), one);
+
+      Block *oBB = &orig->getRegion(1).front();
+      auto term = oBB->getTerminator();
+
+      int revIdx = 1;
+      for (auto &&[active, operand] :
+           llvm::zip(operandsActive, term->getOperands())) {
+        if (active) {
+          // Set diffe here, not add because it should not accumulate across
+          // iterations. Instead the new gradient for this operand is passed in
+          // the return of the reverse while body.
+          gutils->setDiffe(operand, body->getArgument(revIdx), bodyBuilder);
+          revIdx++;
+        }
+      }
+
+      auto first = oBB->rbegin();
+      first++; // skip terminator
+
+      auto last = oBB->rend();
+
+      for (auto it = first; it != last; ++it) {
+        Operation *op = &*it;
+        valid &= gutils->Logic.visitChild(op, bodyBuilder, gutils).succeeded();
+      }
+
+      SmallVector<Value> newResults;
+      newResults.reserve(operands.size());
+      newResults.push_back(bodyIterVar);
+
+      for (auto &&[active, arg] :
+           llvm::zip(operandsActive, oBB->getArguments())) {
+        if (active) {
+          newResults.push_back(gutils->diffe(arg, bodyBuilder));
+          if (!gutils->isConstantValue(arg))
+            gutils->zeroDiffe(arg, bodyBuilder);
+        }
+      }
+
+      bodyBuilder.create<ReturnOp>(orig->getLoc(), newResults);
+    }
+
+    int revIdx = 1;
+    for (auto &&[active, arg] :
+         llvm::zip(operandsActive, orig->getOperands())) {
+      if (active) {
+        if (!gutils->isConstantValue(arg))
+          gutils->addToDiffe(arg, revWhile->getResult(revIdx), builder);
+        revIdx++;
+      }
+    }
+
+    return success(valid);
+  }
+
+  SmallVector<Value> cacheValues(Operation *orig,
+                                 MGradientUtilsReverse *gutils) const {
+    // The primal is augmented to store the number of iterations
+
+    auto newWhile = cast<WhileOp>(gutils->getNewFromOriginal(orig));
+    auto cond = &newWhile.getCond().front();
```

Comment:
Usually one of the variables is a loop induction variable -- which means we don't need to store this. Not sure if this optimization would go here or in a canonicalization though.

Way more powerful versions of this would be here: https://github.com/llvm/Polygeist/blob/77c04bb2a7a2406ca9480bcc9e729b07d2c8d077/lib/polygeist/Passes/CanonicalizeFor.cpp#L662

The pro of doing this early though, is that we can potentially determine a static loop count and then make all of the augmented tensors <Loopsizexf32> instead of <?x32>

Reply:
I was able to reuse the analysis from enzyme-hlo-unroll which we can make more robust in the future to support more complex patterns.

Reply:
in this case I don't think we even care if the value is a constant, we just cache that value regardless of constant or not

Reply:
[and if it's constant cache fixups ought clean it up automatically for us]